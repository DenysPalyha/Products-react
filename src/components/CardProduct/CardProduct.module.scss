.container {
  width: 260px;
  padding: 10px;
  margin: 0 20px 30px 0;
  border: 1px solid #959595;
  border-radius: 5px;
}

.container:nth-child(4n) {
  margin-right: 0;
}

.container:hover {
  box-shadow: 0 0 10px #959595;
}

.wrapperHeader {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.productName {
  margin: 0;
  padding: 0;
  font-size: 16px;
  line-height: 1;
  color: #959595;
}

.starIcon {
  fill: #ffffff;
  width: 24px;
  height: 24px;
}

.starIconActive {
  fill: #f8c660;
  width: 24px;
  height: 24px;
}

.positionImg {
  margin: 10px 0 10px;
  text-align: center;
}

.imagesCard {
  width: 150px;
  height: 200px;
}

.priceCard {
  margin: 0;
  padding: 0 0 10px;
  font-size: 14px;
  line-height: 1;
  color: #959595;
}

.codeProduct {
  margin: 0;
  padding: 0 0 10px;
  font-size: 14px;
  line-height: 1;
  color: #959595;
}

.textColorProduct {
  display: flex;
  font-size: 16px;
  line-height: 1;
  color: #959595;
}

.colorProduct {
  margin: 2px 0 20px 10px;
  width: 15px;
  height: 15px;
  border-radius: 50%;
}

.positionBtn {
  text-align: center;
}

.containerList {
  margin-bottom: 20px;
  padding: 10px;
  width: 100%;
  border: 1px solid #959595;
  border-radius: 5px;

  display: flex;
  justify-content: space-around;
}

.imagesCardList {
  width: 100px;
  height: 100px;
}

.wrapperTextInfoList {
  width: 600px;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.productNameList {
  margin: 0;
  padding: 0;
  font-size: 18px;
  line-height: 1;
  color: #959595;
}

.priceCardList {
  margin: 0;
  padding: 0;
  font-size: 18px;
  line-height: 1;
  color: #959595;
}

.codeProductList {
  margin: 0;
  padding: 0;
  font-size: 18px;
  line-height: 1;
  color: #959595;
}

.textColorProductList {
  display: flex;
  font-size: 18px;
  line-height: 1;
  color: #959595;
}

.colorProductList {
  margin: 2px 0 0 10px;
  width: 15px;
  height: 15px;
  border-radius: 50%;
}

.wrapperHeaderList {
  width: 200px;
  display: flex;
  flex-direction: row-reverse;
  justify-content: space-between;
  align-items: center;
}
