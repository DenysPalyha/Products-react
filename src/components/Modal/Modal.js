import React from "react";
import PropTypes from "prop-types";
import styles from "./Modal.module.scss";

const Modal = ({ header, text, closeButton, actions, closeModale }) => {
  return (
    <div className={styles.modal} data-testid="modal-root">
      <div className={styles.bgColorModal} onClick={closeModale}></div>
      <div className={styles.container}>
        <div className={styles.headerModal}>
          <h2 className={styles.headerTitle}>{header}</h2>
          {closeButton && (
            <button className={styles.btnClose} onClick={closeModale}></button>
          )}
        </div>
        <div className={styles.modalBody}>{text}</div>
        <div className={styles.footerBtn}>{actions}</div>
      </div>
    </div>
  );
};

Modal.propTypes = {
  header: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  actions: PropTypes.node.isRequired,
  closeModale: PropTypes.func.isRequired,
  closeButton: PropTypes.bool,
};

Modal.defaultProps = {
  closeButton: true,
};

export default Modal;
