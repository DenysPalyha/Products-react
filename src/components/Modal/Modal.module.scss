.modal {
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  z-index: 100;
  display: flex;
  justify-content: center;
  align-items: center;
}

.bgColorModal {
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1;
  cursor: pointer;
}

.container {
  width: 500px;
  position: relative;
  border-radius: 5px;
  z-index: 2;
  background: #e74c3c;
}

.headerModal {
  padding: 20px 30px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  background: rgba(0, 0, 0, 0.1);
}

.headerTitle {
  margin: 0;
  font-size: 20px;
  line-height: 1;
  color: #fff;
  font-family: sans-serif;
}

.btnClose {
  width: 30px;
  height: 30px;
  position: relative;
  margin: 0;
  padding: 0;
  outline: none;
  border: none;
  background: transparent;
  cursor: pointer;

  &::after,
  &::before {
    content: "";
    position: absolute;
    height: 1px;
    width: 100%;
    top: 50%;
    left: 0;
    background: rgb(255, 255, 255);
    border-radius: 5px;
  }

  &::after {
    transform: (rotate(45deg));
  }
  &::before {
    transform: (rotate(-45deg));
  }
}

.modalBody {
  padding: 30px 40px;
  font-size: 18px;
  line-height: 2;
  color: #fff;
  text-align: center;
}

.footerBtn {
  text-align: center;
  margin-bottom: 20px;
}
