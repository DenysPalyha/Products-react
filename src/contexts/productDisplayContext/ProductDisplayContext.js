import { createContext } from "react";

const ProductDisplayContext = createContext();

export default ProductDisplayContext;
