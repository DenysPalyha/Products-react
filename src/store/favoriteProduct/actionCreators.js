import {
  SET_FAVORITE_PRODUCT,
  GET_FAVORITE_PRODUCTS,
  COUNT_FAVORITE_PRODUCT,
} from "./actions";

export const setFavoriteProductAC = (payload) => {
  return { type: SET_FAVORITE_PRODUCT, payload };
};

export const getFevoriteProducts = () => (dispatch) => {
  const favorites = JSON.parse(localStorage.getItem("favorite"));
  const countFavorite = JSON.parse(localStorage.getItem("countFavorite"));

  if (favorites !== null) {
    dispatch({ type: GET_FAVORITE_PRODUCTS, payload: favorites });
  }
  if (countFavorite !== null) {
    dispatch({ type: COUNT_FAVORITE_PRODUCT, payload: countFavorite });
  }
};

export const countFavoriteProductAC = () => {
  return { type: COUNT_FAVORITE_PRODUCT };
};
