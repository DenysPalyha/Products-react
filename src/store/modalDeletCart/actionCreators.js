import { SET_MODAL_DELETE_PRODUCT, DATA_PRODUCT_CART } from "./actions";

export const setMoadalDeleteProductAC = (payload) => ({
  type: SET_MODAL_DELETE_PRODUCT,
  payload,
});

export const dataProductCartAC = (payload) => ({
  type: DATA_PRODUCT_CART,
  payload,
});
