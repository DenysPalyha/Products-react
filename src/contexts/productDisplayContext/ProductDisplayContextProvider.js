import { useState } from "react";

import ProductDisplayContext from "./ProductDisplayContext";

function ProductDisplayContextProvider(props) {
  const [isProductDisplay, setIsProductDisplay] = useState(true);
  const { children } = props;

  return (
    <ProductDisplayContext.Provider
      value={{ isProductDisplay, setIsProductDisplay }}
    >
      {children}
    </ProductDisplayContext.Provider>
  );
}

export default ProductDisplayContextProvider;
