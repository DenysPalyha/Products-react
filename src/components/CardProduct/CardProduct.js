import React, { useContext } from "react";
import { useSelector, useDispatch } from "react-redux";
import PropTypes from "prop-types";
import { ReactComponent as StarIcon } from "../../img/icon/star.svg";
import Button from "../Button/Button";

import {
  setMoadalAddProducts,
  dataProduct,
} from "../../store/modalProducts/actionCreators";
import {
  setFavoriteProductAC,
  countFavoriteProductAC,
} from "../../store/favoriteProduct/actionCreators";
import { setIsFavoriteProductAC } from "../../store/products/actionCreators";
import { findAndSetIsFaforiteInCart } from "../../store/productCart/actionCreators";

import styles from "./CardProduct.module.scss";
import ProductDisplayContext from "../../contexts/productDisplayContext/ProductDisplayContext";

const CardProduct = ({
  id,
  productName,
  isActiveFavorite,
  price,
  imageUrl,
  vendorCode,
  color,
}) => {
  const favorites = useSelector((store) => store.favorites.favorite);

  const { isProductDisplay } = useContext(ProductDisplayContext);

  const dispatch = useDispatch();

  return (
    <>
      {isProductDisplay ? (
        <div className={styles.container} data-testid="card-root">
          <div className={styles.wrapperHeader}>
            <p className={styles.productName}>{productName}</p>
            <Button
              textBtn={
                <StarIcon
                  className={
                    isActiveFavorite ? styles.starIconActive : styles.starIcon
                  }
                />
              }
              hendlerClik={() => {
                dispatch(
                  setFavoriteProductAC({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
                dispatch(setIsFavoriteProductAC({ id, favorites }));
                dispatch(findAndSetIsFaforiteInCart({ id, favorites }));
                dispatch(countFavoriteProductAC());
              }}
            />
          </div>
          <div className={styles.positionImg}>
            <img
              className={styles.imagesCard}
              src={imageUrl}
              alt={productName}
            ></img>
          </div>
          <p className={styles.priceCard}>Price: {price}$</p>
          <p className={styles.codeProduct}>Vendor-code: {vendorCode}</p>
          <div className={styles.textColorProduct}>
            Color:
            <div
              className={styles.colorProduct}
              style={{ backgroundColor: color }}
            ></div>
          </div>
          <div className={styles.positionBtn}>
            <Button
              bgColor="#067305"
              textBtn="Add to cart"
              hendlerClik={() => {
                dispatch(setMoadalAddProducts(true));
                dispatch(
                  dataProduct({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
              }}
            />
          </div>
        </div>
      ) : (
        <div className={styles.containerList} data-testid="card-root-list">
          <div className={styles.positionImg}>
            <img
              className={styles.imagesCardList}
              src={imageUrl}
              alt={productName}
            ></img>
          </div>

          <div className={styles.wrapperTextInfoList}>
            <p className={styles.productNameList}>{productName}</p>
            <p className={styles.priceCardList}>Price: {price}$</p>
            <p className={styles.codeProductList}>Vendor-code: {vendorCode}</p>
            <div className={styles.textColorProductList}>
              Color:
              <div
                className={styles.colorProductList}
                style={{ backgroundColor: color }}
              ></div>
            </div>
          </div>

          <div className={styles.wrapperHeaderList}>
            <Button
              textBtn={
                <StarIcon
                  className={
                    isActiveFavorite ? styles.starIconActive : styles.starIcon
                  }
                />
              }
              hendlerClik={() => {
                dispatch(
                  setFavoriteProductAC({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
                dispatch(setIsFavoriteProductAC({ id, favorites }));
                dispatch(findAndSetIsFaforiteInCart({ id, favorites }));
                dispatch(countFavoriteProductAC());
              }}
            />

            <Button
              bgColor="#067305"
              textBtn="Add to cart"
              hendlerClik={() => {
                dispatch(setMoadalAddProducts(true));
                dispatch(
                  dataProduct({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
              }}
            />
          </div>
        </div>
      )}
    </>
  );
};

CardProduct.propTypes = {
  id: PropTypes.number.isRequired,
  productName: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  imageUrl: PropTypes.string,
  vendorCode: PropTypes.number,
  color: PropTypes.string,
  isActiveFavorite: PropTypes.bool,
  favoriteCardProduct: PropTypes.func,
  addCardInCart: PropTypes.func,
};

CardProduct.defaultProps = {
  imageUrl: "./img/no_icon.png",
  vendorCode: 0,
  color: "#fff",
  favoriteCardProduct: () => {},
  addCardInCart: () => {},
  isActiveFavorite: false,
};

export default CardProduct;
