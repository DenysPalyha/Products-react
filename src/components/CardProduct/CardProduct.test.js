import { render, screen, fireEvent, getByTestId } from "@testing-library/react";
import { Provider, useDispatch } from "react-redux";
import { useContext } from "react";
import { store } from "../../store/index";
import ProductDisplayContextProvider from "../../contexts/productDisplayContext/ProductDisplayContextProvider";

import { setMoadalAddProducts } from "../../store/modalProducts/actionCreators";

import {
  setFavoriteProductAC,
  countFavoriteProductAC,
} from "../../store/favoriteProduct/actionCreators";
import { findAndSetIsFaforiteInCart } from "../../store/productCart/actionCreators";
import ProductDisplayContext from "../../contexts/productDisplayContext/ProductDisplayContext";

import Modal from "../Modal/Modal";
import CardProduct from "./CardProduct";

const Container = () => {
  const dispatch = useDispatch();
  return (
    <>
      <button
        type="button"
        onClick={() => {
          dispatch(setMoadalAddProducts(true));
        }}
      >
        Add to cart
      </button>
      <Modal
        header="modal test"
        text="text test"
        closeButton={true}
        actions={<button>click test</button>}
        closeModale={() => {}}
      />
    </>
  );
};

const MockedProvider = ({ children }) => (
  <ProductDisplayContextProvider>
    <Provider store={store}>
      <Container />
    </Provider>
  </ProductDisplayContextProvider>
);

describe("check the opening of the modal by clicking on the button in the product card", () => {
  test("should open modal window on button click Add to cart", () => {
    render(<MockedProvider />);

    const btn = screen.getByText("Add to cart");
    fireEvent.click(btn);

    expect(screen.getByTestId("modal-root")).toBeInTheDocument();
  });
});

const MockedProviderCardProduct = ({ children }) => (
  <ProductDisplayContextProvider>
    <Provider store={store}>
      <CardProduct id={1} productName="Moto G Stylus 5G" price={739.99} />
    </Provider>
  </ProductDisplayContextProvider>
);

describe("Card Product component", () => {
  test("should do snapshot Card Product component", () => {
    const { asFragment } = render(<MockedProviderCardProduct />);

    expect(asFragment()).toMatchSnapshot();
  });
});

const ContainerBtnFavorite = () => {
  const dispatch = useDispatch();

  return (
    <>
      <button
        type="button"
        onClick={() => {
          dispatch(
            setFavoriteProductAC({
              id: 1,
              productName: "Moto G Stylus 5G",
              isActiveFavorite: false,
              price: 739.99,
              imageUrl: "./img/motorola.jpg",
              vendorCode: 225739,
              color: "#366735",
            })
          );
          dispatch(
            findAndSetIsFaforiteInCart({
              id: 1,
              favorites: [
                {
                  id: 1,
                  productName: "Moto G Stylus 5G",
                  price: 739.99,
                  imageUrl: "./img/motorola.jpg",
                  vendorCode: 225739,
                  color: "#366735",
                },
              ],
            })
          );
          dispatch(countFavoriteProductAC());
        }}
      >
        Add to favorite
      </button>

      <CardProduct
        id={1}
        productName="Moto G Stylus 5G"
        price={739.99}
        isActiveFavorite={true}
        color="#366735"
        imageUrl="./img/motorola.jpg"
        vendorCode={225739}
      />
    </>
  );
};

const ContainerBtnFavoriteDel = () => {
  const dispatch = useDispatch();

  return (
    <>
      <button
        type="button"
        onClick={() => {
          dispatch(
            setFavoriteProductAC({
              id: 1,
              productName: "Moto G Stylus 5G",
              isActiveFavorite: false,
              price: 739.99,
              imageUrl: "./img/motorola.jpg",
              vendorCode: 225739,
              color: "#366735",
            })
          );

          dispatch(
            findAndSetIsFaforiteInCart({
              id: 1,
              favorites: [
                {
                  id: 1,
                  productName: "Moto G Stylus 5G",
                  price: 739.99,
                  imageUrl: "./img/motorola.jpg",
                  vendorCode: 225739,
                  color: "#366735",
                },
              ],
            })
          );
          dispatch(countFavoriteProductAC());
        }}
      >
        Add to favorite
      </button>
    </>
  );
};

const MockedProviderBtnFavorite = ({ children }) => (
  <ProductDisplayContextProvider>
    <Provider store={store}>
      <ContainerBtnFavorite />
    </Provider>
  </ProductDisplayContextProvider>
);

const MockedProviderBtnFavoriteDel = ({ children }) => (
  <ProductDisplayContextProvider>
    <Provider store={store}>
      <ContainerBtnFavoriteDel />
    </Provider>
  </ProductDisplayContextProvider>
);

describe("add and delete favorite card product", () => {
  test("should add and show card product in favorite page", () => {
    render(<MockedProviderBtnFavorite />);

    const btn = screen.getByText("Add to favorite");
    fireEvent.click(btn);

    expect(screen.getByTestId("card-root")).toBeInTheDocument();
  });

  test("should delet card product in favorite page", () => {
    render(<MockedProviderBtnFavoriteDel />);

    const btn = screen.getByText("Add to favorite");
    fireEvent.click(btn);

    expect(screen.queryByTestId("card-root")).not.toBeInTheDocument();
  });
});

const ContainerListLayout = () => {
  const { setIsProductDisplay } = useContext(ProductDisplayContext);

  return (
    <>
      <button
        type="button"
        onClick={() => {
          setIsProductDisplay(false);
        }}
      >
        list
      </button>
      <CardProduct id={1} productName="Moto G Stylus 5G" price={739.99} />
    </>
  );
};

const MockedProviderListLayout = ({ children }) => (
  <ProductDisplayContextProvider>
    <Provider store={store}>
      <ContainerListLayout />
    </Provider>
  </ProductDisplayContextProvider>
);

describe("change layout card product", () => {
  test("should change layout card product on list", () => {
    render(<MockedProviderListLayout />);

    const btnList = screen.getByText("list");
    fireEvent.click(btnList);

    expect(screen.getByTestId("card-root-list")).toBeInTheDocument();
  });

  test("should change layout card product on list snapShot", () => {
    const { asFragment } = render(<MockedProviderListLayout />);

    expect(asFragment()).toMatchSnapshot();
  });
});
