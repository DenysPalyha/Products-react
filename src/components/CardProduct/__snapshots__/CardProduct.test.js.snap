// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Product component should do snapshot Card Product component 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="card-root"
  >
    <div
      class="wrapperHeader"
    >
      <p
        class="productName"
      >
        Moto G Stylus 5G
      </p>
      <button
        class="btn"
        type="button"
      >
        <svg
          class="starIcon"
        >
          star.svg
        </svg>
      </button>
    </div>
    <div
      class="positionImg"
    >
      <img
        alt="Moto G Stylus 5G"
        class="imagesCard"
        src="./img/no_icon.png"
      />
    </div>
    <p
      class="priceCard"
    >
      Price: 739.99$
    </p>
    <p
      class="codeProduct"
    >
      Vendor-code: 0
    </p>
    <div
      class="textColorProduct"
    >
      Color:
      <div
        class="colorProduct"
        style="background-color: rgb(255, 255, 255);"
      />
    </div>
    <div
      class="positionBtn"
    >
      <button
        class="btn"
        style="background-color: rgb(6, 115, 5);"
        type="button"
      >
        Add to cart
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`change layout card product should change layout card product on list snapShot 1`] = `
<DocumentFragment>
  <button
    type="button"
  >
    list
  </button>
  <div
    class="container"
    data-testid="card-root"
  >
    <div
      class="wrapperHeader"
    >
      <p
        class="productName"
      >
        Moto G Stylus 5G
      </p>
      <button
        class="btn"
        type="button"
      >
        <svg
          class="starIcon"
        >
          star.svg
        </svg>
      </button>
    </div>
    <div
      class="positionImg"
    >
      <img
        alt="Moto G Stylus 5G"
        class="imagesCard"
        src="./img/no_icon.png"
      />
    </div>
    <p
      class="priceCard"
    >
      Price: 739.99$
    </p>
    <p
      class="codeProduct"
    >
      Vendor-code: 0
    </p>
    <div
      class="textColorProduct"
    >
      Color:
      <div
        class="colorProduct"
        style="background-color: rgb(255, 255, 255);"
      />
    </div>
    <div
      class="positionBtn"
    >
      <button
        class="btn"
        style="background-color: rgb(6, 115, 5);"
        type="button"
      >
        Add to cart
      </button>
    </div>
  </div>
</DocumentFragment>
`;
